validator/client/runner.go

```diff
@@ -64,37 +69,80 @@ func run(ctx context.Context, v Validator) {
 			log.Fatalf("Slasher is not ready: %v", err)
 		}
 	}
-	if err := v.WaitForChainStart(ctx); err != nil {
-		log.Fatalf("Could not determine if beacon chain started: %v", err)
-	}
-	if err := v.WaitForSync(ctx); err != nil {
-		log.Fatalf("Could not determine if beacon node synced: %v", err)
-	}
-
+	ticker := time.NewTicker(backOffPeriod)
+	var headSlot uint64
+	defer ticker.Stop()
 	accountsChangedChan := make(chan struct{}, 1)
-	go handleAccountsChanged(ctx, v, accountsChangedChan)
-	if err := v.WaitForActivation(ctx, accountsChangedChan); err != nil {
-		log.Fatalf("Could not wait for validator activation: %v", err)
+	for {
+		var done bool
+		err := v.WaitForChainStart(ctx)
+		if err != nil && errors.Is(err, errConnectionIssue) {
+			log.Warnf("Could not determine if beacon chain started: %v", err)
+			continue
+		}
+		if err != nil {
+			log.Fatalf("Could not determine if beacon chain started: %v", err)
+		}
+		err = v.WaitForSync(ctx)
+		if err != nil && errors.Is(err, errConnectionIssue) {
+			log.Warnf("Could not determine if beacon chain started: %v", err)
+			continue
+		}
+		if err != nil {
+			log.Fatalf("Could not determine if beacon node synced: %v", err)
+		}
+		err = v.WaitForActivation(ctx, accountsChangedChan)
+		if err != nil && errors.Is(err, errConnectionIssue) {
+			log.Warnf("Could not wait for validator activation: %v", err)
+			continue
+		}
+		if err != nil {
+			log.Fatalf("Could not wait for validator activation: %v", err)
+		}
+		headSlot, err = v.CanonicalHeadSlot(ctx)
+		if err != nil && errors.Is(err, errConnectionIssue) {
+			log.Warnf("Could not get current canonical head slot: %v", err)
+			continue
+		}
+		if err != nil {
+			log.Fatalf("Could not get current canonical head slot: %v", err)
+		}
+		done = true
+		select {
```

Comment:
Do we really need this select? Can't we just add `ctx.Err()` check at the top?
```
if ctx.Err() != nil {
		return nil, ctx.Err()
	}
``` 
and break once we get to the end of the loop

I think that is simpler to reason?

Reply:
tried to make it clearer